src/ws.nim

```diff
@@ -0,0 +1,570 @@
+  else:
+    b1 = 127u8
+
+  if f.mask:
+    b1 = b1 or (1 shl 7)
+
+  ret.write(uint8 b1)
+
+  # Only need more bytes if data len is 7+16 bits, or 7+64 bits.
+  if f.data.len > 125 and f.data.len <= 0xffff:
+    # Data len is 7+16 bits.
+    ret.write(htons(f.data.len.uint16))
+  elif f.data.len > 0xffff:
+    # Data len is 7+64 bits.
+    var len = f.data.len
+    ret.write char((len shr 56) and 255)
+    ret.write char((len shr 48) and 255)
+    ret.write char((len shr 40) and 255)
+    ret.write char((len shr 32) and 255)
+    ret.write char((len shr 24) and 255)
+    ret.write char((len shr 16) and 255)
+    ret.write char((len shr 8) and 255)
+    ret.write char(len and 255)
+
+  var data = f.data
+
+  if f.mask:
+    # If we need to mask it generate random mask key and mask the data.
+    let maskKey = genMaskKey()
+    for i in 0..<data.len:
+      data[i] = (data[i].uint8 xor maskKey[i mod 4].uint8).char
+    # Write mask key next.
+    ret.write(maskKey)
+
+  # Write the data.
+  ret.write(data)
+  ret.setPosition(0)
+  return ret.readAll()
+
+proc send*(ws: WebSocket, text: string, opcode = Opcode.Text): Future[void] {.async.} =
+  try:
+    var frame = encodeFrame((
+      fin: true,
+      rsv1: false,
+      rsv2: false,
+      rsv3: false,
+      opcode: opcode,
+      mask: ws.masked,
+      data: text
+    ))
+    const maxSize = 1024*1024
+    # Send stuff in 1 megabyte chunks to prevent IOErrors.
+    # This really large packets.
+    var i = 0
+    while i < frame.len:
+      let data = frame[i ..< min(frame.len, i + maxSize)]
+      discard await ws.tcpSocket.write(data)
+      i += maxSize
+      await sleepAsync(1)
+  except Defect, IOError, OSError, ValueError:
+    # Wrap all exceptions in a WebSocketError so its easy to catch
+    raise newException(WebSocketError, "Failed to send data: " &
+        getCurrentExceptionMsg())
+
+proc close*(ws: WebSocket) =
+  ## Close the Socket, sends close packet.
+  ws.readyState = Closed
+  proc close() {.async.} =
+    await ws.send("", Close)
+    ws.tcpSocket.close()
+  asyncCheck close()
+
+proc toString(bytes: seq[byte]): string =
+  result = ""
+  for byte in bytes: result.add(char(byte))
+
+proc receiveFrame(ws: WebSocket): Future[Frame] {.async.} =
+  ## Gets a frame from the WebSocket.
+  ## See https://tools.ietf.org/html/rfc6455#section-5.2
+
+  if cast[int](ws.tcpSocket.fd) == -1:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+
+  # Grab the header.
+  var header: seq[byte]
+  try:
+    header = await ws.tcpSocket.read(2)
+  except:
+    raise newException(WebSocketError, "Socket closed")
+
+  if header.len != 2:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+
+  let b0 = header[0].uint8
+  let b1 = header[1].uint8
+
+  # Read the flags and fin from the header.
+  result.fin = b0[0]
+  result.rsv1 = b0[1]
+  result.rsv2 = b0[2]
+  result.rsv3 = b0[3]
+  result.opcode = (b0 and 0x0f).Opcode
```

Comment:
`Opcode` is enum with holes, such conversion could lead to exception if `b0` is in `[3, 7]` or `[b, f]`.